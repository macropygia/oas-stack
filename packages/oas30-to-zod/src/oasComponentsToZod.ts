import path from 'node:path'
// import { fileURLToPath } from 'node:url'

import ejs from 'ejs'
import fse from 'fs-extra'
import SwaggerParser from '@apidevtools/swagger-parser'
import { red } from 'ansis/colors'
import type { OpenAPI, OpenAPIV3 } from 'openapi-types'

import { defaultParseOptions, parseSchema } from '@/parseSchema.js'
import type {
  Document,
  ComponentGraph,
  Options,
  ComponentName,
} from '@/types/index.js'
import { format } from '@/utils/format.js'
import { generatePrintingOrder } from '@/utils/generatePrintingOrder.js'
import { autocompleteObject } from '@/utils/autocompleteObject.js'
import { isV3, isValidDoc } from '@/utils/validateDoc.js'

// const __filename = fileURLToPath(import.meta.url)
// const __dirname = path.dirname(__filename)

const defaultOptions: Options = {
  output: false,
  dereference: false,
  disableAutocomplete: false,
  exportName: 'schemas',
  individually: false,
  eslintDisable: false,
  disableRules: [],
  withoutImport: false,
  withoutExport: false,
  inheritPrettier: false,
  disableFormat: false,
  ...defaultParseOptions,
}

/**
 * Generate Zod schemas from OpenAPI 3.0 Components Object
 * @param input - Document object or file path
 * @param userOptions - Options
 * @returns Zod schemas (JavaScript)
 */
export const oasComponentsToZod = async (
  input: OpenAPIV3.Document | string,
  userOptions?: Options
): Promise<string> => {
  const options: Options = { ...defaultOptions, ...userOptions }

  const doc = await init(input, options)

  const dereferencedDoc = await SwaggerParser.dereference(
    JSON.parse(JSON.stringify(doc))
  ).then((deref) => deref as Document)

  const graph = generateGraph(dereferencedDoc)

  const targetDoc = options.dereference ? dereferencedDoc : doc
  const targetSchema = targetDoc.components.schemas

  // Reserved object for cusomization
  const data: Record<string, any> = {}

  const parsedComponents: Record<ComponentName, string> = {}
  Object.entries(targetSchema).forEach(([compName, compSchema]) => {
    const deps: ComponentName[] = []
    parsedComponents[compName] = parseSchema(compSchema, {
      options,
      parsers: options.parsers,
      graph,
      name: compName,
      deps,
      doc: targetDoc,
      schemas: targetDoc.components.schemas,
      data,
    })
    graph.deps[compName] = deps
  })

  // Create a shallow copy of deps for destructive processing
  const componentDeps = Object.assign({}, graph.deps)
  const printingOrder = generatePrintingOrder(componentDeps)

  const template = options.template
    ? fse.readFileSync(options.template).toString()
    : fse.readFileSync(path.resolve(__dirname, 'default.ts.ejs')).toString()

  const rawParsed = ejs.render(template, {
    // Processed data
    parsedComponents,
    printingOrder,
    // Options
    ...options,
    // Graph
    graph,
    data,
  })

  const parsed = options.disableFormat
    ? rawParsed
    : format(rawParsed, options.inheritPrettier)

  if (options.output) fse.outputFileSync(options.output, parsed)

  return parsed
}

/**
 * Verify settings and document
 * @param input - Document object or file path
 * @param options - Options
 * @returns Document object
 */
const init = async (input: OpenAPI.Document | string, options: Options) => {
  if (options.output)
    await fse.ensureFile(options.output).catch((err) => {
      console.error(red`Unable to create output file.`)
      throw new Error(err)
    })

  if (options.template && !fse.existsSync(options.template))
    throw new Error(`EJS template does not exists: ${options.template}`)

  const doc =
    typeof input === 'string' ? await SwaggerParser.bundle(input) : input

  if (!isV3(doc)) {
    console.error(
      red`Document version mismatch. Only version 3.0 is supported.`
    )
    throw new Error('Only version 3.0 is supported.')
  }

  if (!isValidDoc(doc))
    throw new Error(`Document has no 'components' or 'schemas'.`)

  if (!options.disableAutocomplete) autocompleteObject(doc)

  return doc
}

/**
 * Generate component graph from dereferenced document
 * @param dereferencedDoc - Dereferenced document
 */
const generateGraph = (dereferencedDoc: Document): ComponentGraph => {
  const compGraph: ComponentGraph = {
    deps: {},
    isObject: {},
    isNullable: {},
    hasDefault: {},
  }

  const resolvedSchemas = (dereferencedDoc as Document).components.schemas

  Object.entries(resolvedSchemas).forEach(([compName, compSchema]) => {
    // Object
    if ('type' in compSchema && compSchema.type === 'object')
      compGraph.isObject[compName] = true
    else compGraph.isObject[compName] = false

    // Nullable
    if ('nullable' in compSchema && compSchema.nullable === true)
      compGraph.isNullable[compName] = true
    else compGraph.isNullable[compName] = false

    // Default
    if ('default' in compSchema) compGraph.hasDefault[compName] = true
    else compGraph.hasDefault[compName] = false
  })

  return compGraph
}
