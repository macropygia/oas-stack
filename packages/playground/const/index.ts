import type { Options } from '../types'

export const defaultPrettier = `{
  "printWidth": 60
}
`

export const defaultOptions: Options = {
  dereference: false,
  exportName: undefined,
  individually: false,
  eslintDisable: false,
  disableRules: undefined,
  withoutImport: false,
  withoutExport: false,
  disableFormat: false,
  withoutDefaults: false,
  withDesc: false,
  withAnchors: false,
  disableAutocomplete: false,
  fontSize: 14,
  inheritPrettier: defaultPrettier,
  splitRatio: 50,
}

export const defaultDoc = `openapi: 3.0.0
info:
  version: 1.0.0
  title: sample
paths:
components:
  schemas:
    Comp1:
      type: string
      minLength: 3
      maxLength: 10
      pattern: '[a-z]+'
      default: foo
      description: bar
    Comp2:
      type: boolean
    Comp3:
      type: integer
    Comp4:
      type: string
      enum:
        - foo
        - bar
      default:
        - foo
    Comp5:
      type: object
      required:
        - Prop1
        - Prop5
      properties:
        Prop1:
          $ref: '#/components/schemas/Comp1'
        Prop2:
          $ref: '#/components/schemas/Comp2'
        Prop3:
          $ref: '#/components/schemas/Comp3'
        Prop4:
          $ref: '#/components/schemas/Comp4'
        Prop5:
          type: number
          minimum: 100
          exclusiveMinimum: true
`
