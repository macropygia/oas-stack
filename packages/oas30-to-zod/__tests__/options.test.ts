import fse from 'fs-extra'
import { describe, test, expect } from 'vitest'

import { oasComponentsToZod } from '@/oasComponentsToZod.js'
import { minimumOutput } from '@tests/const.js'

describe('Options', async () => {
  test('Disable format', async () => {
    await expect(
      oasComponentsToZod('__tests__/minimum.yml', {
        disableFormat: true,
      })
    ).resolves.toMatchInlineSnapshot(`
      "import { z } from 'zod';

      const Minimum = z.string();

      export const schemas = {
        Minimum
      };
      "
    `)
  })

  test('Enable output', async () => {
    await oasComponentsToZod('__tests__/minimum.yml', {
      ...minimumOutput,
      output: '__tests__/minimum.ts',
    })

    expect(fse.readFileSync('__tests__/minimum.ts').toString())
      .toMatchInlineSnapshot(`
      "const Minimum = z.string();
      "
    `)

    fse.removeSync('__tests__/minimum.ts')
  })

  test('Enable dereference', async () => {
    await expect(
      oasComponentsToZod('__tests__/dereference.yml', {
        ...minimumOutput,
        dereference: true,
      })
    ).resolves.toMatchSnapshot()
  })

  test('Change export variable name', async () => {
    await expect(
      oasComponentsToZod('__tests__/minimum.yml', {
        withoutImport: true,
        exportName: 'foo',
      })
    ).resolves.toMatchInlineSnapshot(`
      "const Minimum = z.string();

      export const foo = {
        Minimum,
      };
      "
    `)
  })

  test('Export individually', async () => {
    await expect(
      oasComponentsToZod('__tests__/minimum.yml', {
        ...minimumOutput,
        individually: true,
      })
    ).resolves.toMatchInlineSnapshot(`
      "export const Minimum = z.string();
      "
    `)
  })

  test('Add `eslint-disable`', async () => {
    await expect(
      oasComponentsToZod('__tests__/minimum.yml', {
        ...minimumOutput,
        eslintDisable: true,
      })
    ).resolves.toMatchInlineSnapshot(`
      "/* eslint-disable */

      const Minimum = z.string();
      "
    `)
  })

  test('Add `eslint-disable` with single rule', async () => {
    await expect(
      oasComponentsToZod('__tests__/minimum.yml', {
        ...minimumOutput,
        eslintDisable: true,
        disableRules: ['no-control-regex'],
      })
    ).resolves.toMatchInlineSnapshot(`
      "/* eslint-disable no-control-regex */

      const Minimum = z.string();
      "
    `)
  })

  test('Add `eslint-disable` with multiple rules', async () => {
    await expect(
      oasComponentsToZod('__tests__/minimum.yml', {
        ...minimumOutput,
        eslintDisable: true,
        disableRules: ['no-control-regex', 'import/no-named-export'],
      })
    ).resolves.toMatchInlineSnapshot(`
      "/* eslint-disable no-control-regex, import/no-named-export */

      const Minimum = z.string();
      "
    `)
  })

  test('Disable `.default()`', async () => {
    await expect(
      oasComponentsToZod('__tests__/string.yml', {
        ...minimumOutput,
        withoutDefaults: true,
      })
    ).resolves.toMatchInlineSnapshot(`
      "const Minimum = z.string().min(1).max(10).regex(new RegExp("[a-z]+"));
      "
    `)
  })

  test('Enable `.descrive()`', async () => {
    await expect(
      oasComponentsToZod('__tests__/string.yml', {
        ...minimumOutput,
        withDesc: true,
      })
    ).resolves.toMatchInlineSnapshot(`
      "const Minimum = z
        .string()
        .min(1)
        .max(10)
        .regex(new RegExp("[a-z]+"))
        .default("foo")
        .describe("bar");
      "
    `)
  })

  test('Use external template', async () => {
    await expect(
      oasComponentsToZod('__tests__/string.yml', {
        disableFormat: true,
        template: '__tests__/test.ts.ejs',
      })
    ).resolves.toMatchInlineSnapshot('"External Template"')
  })
})
