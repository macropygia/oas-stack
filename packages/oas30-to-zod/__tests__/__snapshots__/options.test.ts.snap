// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Options > Enable dereference 1`] = `
"const Comp1 = z.object({ Prop1: z.string(), Prop2: z.number() }).partial();
const Comp2 = z.object({ Prop1: z.string(), Prop2: z.number() }).partial();
"
`;
